### ISSUE TITLE
$.type

### ISSUE THREAD

**Issue by Aiosa** [openseadragon/openseadragon#2237]
I noticed incorrect behavior of `OpenSeadragon.type` function against the newer ES API. If you define `async function x(...){`
the console will output:
```
OpenSeadragon.isFunction(x)
  false
OpenSeadragon.type(x)
  'object'
typeof x
  'function'
toString.call(x)
  '[object AsyncFunction]'
```
IMHO `OpenSeadragon.type(x)` should return `function` as well. Maybe it could be worth the effort to try to fix the behavior of built-in utilities against new JS features...

**Comment by iangilman** [comment]
Good catch! Thank you for filing this :)

**Comment by Aiosa** [comment]
I actually arrived at this issue once more, since I tried to re-use `EventSource` in the application I am building. Creating `export-data` events that work with async functions since it's IO after all, `addHandler` on `async function` exists with no-op since `OpenSeadragon.isFunction()` evaluates to false. Furthermore, I would probably want OSD to wait for async handlers, or have an option that would enable waiting. I will try creating PR. The only thing I am unsure about is how to approach ES6 features support while _not_ using ES6 syntax... Time to upgrade? I mean, it's 2022 and libraries such as [fabricJS](https://github.com/fabricjs/fabric.js/issues/8299) tend to shift toward promises API.

**Comment by iangilman** [comment]
Yeah, good question! We still haven't officially sunset IE11 support (but maybe it's time to consider it?). I don't know how much ES6 IE11 supports. 

Either way, I'm not really in favor of doing a massive overhaul of the entire codebase to bring it up to ES6. Let's just add the bits we need as we need them. 

So I guess my suggestion is to make the PR, and we can evaluate the necessary ES6 on a case by case basis. Sound fair?

**Comment by Aiosa** [comment]
So I actually managed to create https://github.com/openseadragon/openseadragon/pull/2273 while not really using much of ES6, the only thing I used were two `Promise` functions, so I added their error proxies as well, which seems like a stable solution. When I was meddling with EventSource I also added support for priorities and canceling events, which I have been thinking about for quite some time. I still have several questions about the implementation I proposed.
